// @nestjs/devtools-integration/dist/devtools.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DevtoolsModule = void 0;
const common_1 = require("@nestjs/common");
const discovery_module_1 = require("@nestjs/core/discovery/discovery-module");
const devtools_http_server_host_1 = require("./devtools-http-server.host");
const devtools_module_definition_1 = require("./devtools.module-definition");
const sandboxed_code_executor_1 = require("./sandboxed-code.executor");
let DevtoolsModule = class DevtoolsModule extends devtools_module_definition_1.ConfigurableModuleClass {
};
DevtoolsModule = __decorate([
    (0, common_1.Module)({
        imports: [discovery_module_1.DiscoveryModule],
        providers: [
            devtools_http_server_host_1.DevtoolsHttpServerHost,
            sandboxed_code_executor_1.SandboxedCodeExecutor,
            {
                provide: devtools_module_definition_1.DEVTOOLS_MODULE_OPTIONS,
                useFactory: () => {
                    throw new Error(`DevtoolsModule should be initialized with either "register" or "registerAsync" method.`);
                },
            },
        ],
        exports: [],
    })
], DevtoolsModule);
exports.DevtoolsModule = DevtoolsModule;
